### PR TITLE
Brian master

### PR DESCRIPTION
**Pull request by wbhart**

### CODE REVIEW

**Comment by wbhart** [comment]
I've added a review above. There are three issues that need to be resolved before I can merge this. Otherwise, it looks good.

**Comment by BrianGladman** [comment]
@wbhart   Sorry to interfere with your work Bill, can I do anything to help with  these three issues?  Do I need to make a request on MPIR-devel for a LInux build guru to advise on adding the mpn\generic\zero_p.c file to the Linux build?

**Comment by wbhart** [comment]
@BrianGladman The purpose of the review is to let you know what problems there are with the PR. So I'm certainly hoping you can help with this.

You should get a copy of my brian_master branch and make the necessary commits there, then let me know so that I can add them to the pull request. Please do not merge into or rebase onto one of your branches. Simply check out my branch (as a tracking branch) in your repository after fetching from me!

If you are unable to figure out how to add the mpn file, then yes, we need to ask on the list. But there's no point asking anyone to do that until I have merged this PR from you. No one is going to want to screw around with all these branches.

File: tests/mpn/Makefile.am [hunk 1]
```diff
@@ -23,7 +23,7 @@
...
 
-check_PROGRAMS = t-addadd_n t-addsub_n t-aors_1 t-asmtype t-dc_bdiv_q t-dc_bdiv_q_n t-dc_bdiv_qr t-dc_bdiv_qr_n t-dc_div_q t-dc_div_qr t-dc_div_qr_n t-dc_divappr_q t-divebyff t-divebyfobm1 t-divrem_1 t-fat t-gcdext t-get_d t-hgcd t-instrument t-inv_div_q t-inv_div_qr t-inv_div_qr_n t-inv_divappr_q t-inv_divappr_q_n t-invert t-iord_u logic t-lorrshift1 t-matrix22 t-mp_bases t-mullow_basecase t-mullowhigh t-mulmid t-mulmod_2expm1 t-mulmod_2expp1 t-neg t-perfsqr t-redc_1 t-sb_bdiv_q t-sb_bdiv_qr t-sb_div_q t-sb_div_qr t-sb_divappr_q t-scan t-sizeinbase t-subadd_n t-tdiv_q t-tdiv_qr 
+check_PROGRAMS = t-addadd_n t-addsub_n t-aors_1 t-asmtype t-dc_bdiv_q t-dc_bdiv_q_n t-dc_bdiv_qr t-dc_bdiv_qr_n t-dc_div_q t-dc_div_qr t-dc_div_qr_n t-dc_divappr_q t-divebyff t-divebyfobm1 t-divrem_1 t-fat t-gcdext t-get_d t-hgcd t-instrument t-inv_div_q t-inv_div_qr t-inv_div_qr_n t-inv_divappr_q t-inv_divappr_q_n t-invert t-iord_u t-logic t-lorrshift1 t-matrix22 t-mp_bases t-mullow_basecase t-mullowhigh t-mulmid t-mulmod_2expm1 t-mulmod_2expp1 t-neg t-perfsqr t-redc_1 t-sb_bdiv_q t-sb_bdiv_qr t-sb_div_q t-sb_div_qr t-sb_divappr_q t-scan t-sizeinbase t-subadd_n t-tdiv_q t-tdiv_qr
```
Comment (BrianGladman) [inline tests/mpn/Makefile.am]: this adds the mpn logic tests to MPIR (we didn't have them) - the brian master branch has this but your master does not

**Comment by wbhart** [comment]
I just merged again from your master, and it has added all 1900 files back
in!

This should not have happened. Where did you put the patch to add the new
mpn file on the linux side?

**Comment by wbhart** [comment]
Oh, I see. You have requested changes to the review on GitHub. I'm wondering how you did that, because it is trying to modify thousands of files again.

**Comment by BrianGladman** [comment]
Oh, I am sorry Bill, I was just trying to use the GITHUB comment/review process but I must admit that it is near useless in this case since, as you commented earlier, it doesn't compare what we need compared.  I will stop doing this as it seems to be causing problems.  My master has changed though - would this cause the problem? I have just put the new msvc build into it.   I can take it out again if this is the problem (it was horrendous when I came to put it back - my msvc_build branch failed completely so I had to delete iit completely and add it as if it had never been there).

**Comment by BrianGladman** [comment]
@wbhart 

I have my Visual Studio set to your brian_master branch and I am going through the PR on-line and then checking in the brian_master branch in Visual Studio to see what might be causing the issue. I am then commenting on it using the GITHUB comment mechanism.   I am not making any actual corrections anywhere as I have quicckly concluded that all of this requires a person with knowledge of the Linux build to sort it out (Jens has kindly agreed to help)

**Comment by BrianGladman** [comment]
Unless GIT is playing up in some way on my system, I can see an issue with brian_master that we will need to fix - it appears to have some but not all of the new msvc build files in it.  This is my problem since it should have been completely empty and is the reason why I had a horrendous experience when I tried to put the new build files back into my master.

**Comment by wbhart** [comment]
@Brian Gladman <brg@gladman.plus.com> Simply delete your brian_master
branch and again checkout mine. Make sure you don't have any uncommitted
changes in your repo before doing this, of course.

I will leave the review open, but I will have to close it, as I cannot
accept the changes as they are. Once you have everything correct in the new
brian_master branch just push it to your GitHub and let me know.

The problem is all these autogenerated files, which tells me we need to
remove them all from the repository for ordinary development. You could
still have them in your local repository, but they would need to be in
.gitignore so they don't get committed.

**Comment by BrianGladman** [comment]
@wbhart  

Having the Visual Studio project files in my local directory won't help people who need them since they won't be in my repository.   

I am wondering if there is an alternative.  Could I put them in a branch of my MPIR repository that never gets merged into my master branch?   That should make them available without messing up inter-repository operations shouldn't it?    

They would have to get them from my repository rather than yours but that is not a problem since people using MPIR on WIndows have been doing this already for the last couple of years.  And it wouldn't mean that I was running a a fork since our two masters would remain in sync.

**Comment by wbhart** [comment]
I don't mind how you deal with it. I doubt that it will be easy to maintain
with two branches for this though. E.g. you may not be able to merge from
master into the branch with all your files without deleting them.

You could simply make tarballs and put them on your website. That's what
I'd do, and then stick with the single master branch + current working
branch (where all your work gets done).

**Comment by BrianGladman** [comment]
The branch would only hold the vcproj files and it wouldn't matter if they got deleted. If this branch got messed up by merging into it from anywhere I would simply delete and recreate it.   

An alternative is for me to add a separate repository, MPIR_VS, to hold these files.

**Comment by wbhart** [comment]
On Thu, 30 Aug 2018 at 18:46, Brian Gladman <notifications@github.com>
wrote:

> The branch would only hold the vcproj files and it wouldn't matter if they
> got deleted. If this branch got messed up by merging into it from anywhere
> I would simply delete and recreate it.
>

You wouldn't be able to have both the normal MPIR branch and the vcproj
branch checked out at the same time. If  you had the vcproj files in the
directory, but ignored by Git, you wouldn't be able to check out the vcproj
branch. It's pretty much not going to work.


> An alternative is for me to add a separate repository, MPIR_VS, to hold
> these files.
>
Same problem.

As they are autogenerated, I don't see why you would even want them in a
Git repository. It's easier to recreate them than to screw around with Git
every time you need them (I presume).

Anyhow, it's up to you. But I don't see any value in a Git repository or
branch for this. A zip file is just as convenient for distribution!

**Comment by BrianGladman** [comment]
Bill, I am now completely lost.

I just don't see how a totally separate repository on GITHUB can have any problems.  All I would put into this are the auto-generated Visual Studio files needed to build MPIR (i.e. the *.vcproj files).   

The relationship between this repository and my MPIR repository would be *entirely* manual.

**Comment by wbhart** [comment]
Brian, technically it would work fine. You could have a directory within
the mpir source tree on your machine that corresponds to this other
repository, and just ensure that this directory is in the .gitignore of the
main MPIR repository.

I'm just saying I don't see the point of putting autogenerated files on
GitHub. It's like putting binaries on GitHub, which people don't usually
do. For example, what license are you going to use? If you use the GPL you
would have to give the source code for generating those files, and whilst
that is no more complicated than putting a link in the README, it's still
just a little odd.

But if it somehow makes things easier for you (I don't see how), then go
ahead.

**Comment by BrianGladman** [comment]
The *.vcproj files have been in the MPIR directory ever since I started working on MPIR a many years ago.   People using Visual Studio have been using them from the MPIR repository (yours or mine) for all this time and this has led to an expectation that they are going to be there.  With VS 2017 I stopped putting assembler builds in the repository and this generated a lot of requests for help (yes, I know that people should read the docs but they very don't).   I just don't want the hassle of all this email that leaving out these files will generate for me (maybe I'll drop this for VS2019!).

**Comment by wbhart** [comment]
We can certainly add a note to the README, which is prominent when you go
to the repository on GitHub.

>

**Pull request closed without merging**